### PR TITLE
[Servlet] File migration

### PR DESCRIPTION
**Pull request by TomiwaOke**
- Form logic for file form to allow for file migrations

- Add on click listeners to google chart in order to show all migrated numbers

- Allow for text files to be downloaded by users

Link to deployed webapp to view changes: https://phone-number-migrator.uc.r.appspot.com/
Text file to use for file migration: [HelloWorld.txt](https://github.com/TomiwaOke/libphonenumber/files/5255411/HelloWorld.txt)

Please go to the website and run a file migration with the above text file using country code "84". Use country code "966" to see what happens when there are no numbers in the file that can be migrated using a given country code.

**NOTE:** using country codes +1,+86,+55 result in errors at the moment becasue the range files are too big.

### CODE REVIEW

File: migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java [hunk 1]
```diff
@@ -48,8 +61,26 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
...
-      String file = req.getParameter("file");
-      countryCode = req.getParameter("fileCountryCode");
+      handleFileMigration(req, resp);
+    }
+  }
+
+  /** Retrieves the form data for the numbers after a file migration and downloads them as a text file. */
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    String fileName = "+" + req.getParameter("countryCode") + "_Migration_ " + req.getParameter("fileName");
```
Comment (gonecoding) [inline migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java]: I believe we should do at least some sanitization on those request parameters before building a filename from them, e. g. exclude certain special characters like `/`, `\`, `.` etc. (not an exhaustive list). Can you check if there's a library for that? :)

File: migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java [hunk 2]
```diff
@@ -80,4 +111,73 @@ public void handleSingleNumberMigration(String number, String countryCode, HttpS
...
+    if (resultList == null) return ImmutableList.of();
+    return resultList.stream().map(MigrationResult::toString).collect(ImmutableList.toImmutableList());
+  }
+
+  public void handleFileMigration(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    try {
+      Map.Entry<String, MigrationJob.MigrationReport> result = performFileMigration(req).entrySet().iterator().next();
+      String fileName = result.getKey();
+      MigrationJob.MigrationReport report = result.getValue();
+      // List converted into a Set to allow for constant time contains() method below
```
Comment (gonecoding) [inline migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java:133]: Nice touch!

File: migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java [hunk 3]
```diff
@@ -80,4 +111,73 @@ public void handleSingleNumberMigration(String number, String countryCode, HttpS
...
+      req.getRequestDispatcher("index.jsp").forward(req, resp);
+    }
+  }
+
+  private ImmutableMap<String, MigrationJob.MigrationReport> performFileMigration(HttpServletRequest req) throws Exception {
+    String countryCode = "";
+    String file = "";
+    String fileName = "";
+    ServletFileUpload upload = new ServletFileUpload();
+    upload.setSizeMax(50000);
```
Comment (gonecoding) [inline migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java]: Make this into a constant?

File: migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java (hunk 1 above)
Comment (brittanyroesch) [reply migrator/migrator-servlet/src/main/java/com/google/phonenumbers/ServletMain.java]: +1

File: migrator/migrator-servlet/src/main/webapp/index.jsp [hunk 4]
```diff
@@ -67,6 +162,19 @@
                   " to the given phone number or the specified number has never been valid.</p>");
           // TODO: add link for users to file bugs
         }
+      } else if (request.getAttribute("fileError") == null && request.getAttribute("fileName") != null) {
+        out.print("<h3>+" + request.getAttribute("fileCountryCode") + " " + request.getAttribute("fileName") + " Migration Report</h3>");
```
Comment (brittanyroesch) [inline migrator/migrator-servlet/src/main/webapp/index.jsp]: Maybe we could put the filename in single quotations and add "for the country code: " +  request.getAttribute("fileCountryCode") at the end?

File: migrator/migrator-servlet/src/main/webapp/index.jsp [hunk 5]
```diff
@@ -67,6 +162,19 @@
...
+        out.print("<p>Below is a chart showing the ratio of numbers from the entered file that were able to be migrated" +
+                " using +" + request.getAttribute("fileCountryCode") + " migration recipes. To understand more," +
```
Comment (brittanyroesch) [inline migrator/migrator-servlet/src/main/webapp/index.jsp]: nit: maybe single quotes around the fileCountryCode here.

File: migrator/migrator-servlet/src/main/webapp/index.jsp [hunk 6]
```diff
@@ -67,6 +162,19 @@
...
+                " select a given segment from the below donut chart.</p>");
```
Comment (brittanyroesch) [inline migrator/migrator-servlet/src/main/webapp/index.jsp]: nit: maybe just "from the chart below" instead

**Review by brittanyroesch — APPROVED** [review]
Just some small comments - really great work Tomiwa!

**Review by gonecoding — APPROVED** [review]

**Comment by gonecoding** [comment]
Thanks for addressing my comments - great work!

**Pull request merged**